### PR TITLE
CR-1105737 : Enhanced the xbutil2 for AIE

### PR DESCRIPTION
**Pull request by saifuddin-xilinx**
**CR**
CR-1105737 : Some enhancement request on xbutil2 related to AIE.

**Issue**
Some enhancement request on xbutil2 related to AIE when we talked to internal users:
  - When there're many cores in the design, it's convenient just to report status of specific core and tiles.
     -- Select certain tiles from aieshim that are of intertest
     -- Select certain cores from aie that are of intertest
  - Able to run continuously to detect AIE status (include core status)

**Solution**
Added the above support. 

_$./xbutil2 examine -r aieshim -e tiles 0,1 --device 0000:00:00.0_
image

_$./xbutil2 examine -r aieshim -e coress 1 --device 0000:00:00.0_
image

_$./xbutil2 examine -r aie -e less --device 0000:00:00.0_
image

The above command will print only status information once. It's user responsibility to run as a watch mode to get continuous status information. 

_$./xbutil2 examine -r aie -e less cores 1 --device 0000:00:00.0_
This to print status information for only core 1

### CODE REVIEW

File: src/runtime_src/core/tools/common/ReportAie.cpp [hunk 1]
```diff
@@ -21,6 +21,13 @@
 #include "core/common/device.h"
 #include <boost/optional/optional.hpp>
 #include <boost/property_tree/json_parser.hpp>
+#include <boost/algorithm/string.hpp>
+#include <boost/algorithm/string/split.hpp>
+#include <boost/algorithm/string/join.hpp>
+#include <boost/lexical_cast.hpp>
```
Comment (stsoe) [inline src/runtime_src/core/tools/common/ReportAie.cpp:27]: No, please do not use boost::lexical cast

File: src/runtime_src/core/tools/common/ReportAie.cpp [hunk 2]
```diff
@@ -21,6 +21,13 @@
 #include "core/common/device.h"
 #include <boost/optional/optional.hpp>
 #include <boost/property_tree/json_parser.hpp>
+#include <boost/algorithm/string.hpp>
+#include <boost/algorithm/string/split.hpp>
+#include <boost/algorithm/string/join.hpp>
+#include <boost/lexical_cast.hpp>
+#include <boost/thread.hpp>
+#include <boost/asio.hpp>
+#include <boost/asio/signal_set.hpp>
```
Comment (stsoe) [inline src/runtime_src/core/tools/common/ReportAie.cpp:30]: Let's not add a dependency on a new boost libraries.  The boost::thread library is replaced with std::thread.  If you are looking for some functionality that appears to be missing in std::thread vs boost::thread, then please reach out.

File: src/runtime_src/core/tools/xbutil2/CMakeLists.txt [hunk 3]
```diff
@@ -82,7 +82,7 @@ target_link_libraries(${XBUTIL2_NAME}
   )
 
 if (NOT WIN32)
-  target_link_libraries(${XBUTIL2_NAME} PRIVATE pthread uuid dl)
+  target_link_libraries(${XBUTIL2_NAME} PRIVATE pthread boost_thread uuid dl)
```
Comment (stsoe) [inline src/runtime_src/core/tools/xbutil2/CMakeLists.txt:85]: We don't want this dependency unless there is absolutely now other way.

**Review by stsoe — CHANGES_REQUESTED** [review]
This PR does is lacking a good commit description that explains why dependencies were added on new boost libraries.   Without a detailed explanation, this request cannot be merged.

File: src/runtime_src/core/tools/common/ReportAie.cpp [hunk 4]
```diff
@@ -502,14 +510,108 @@ ReportAie::getPropertyTree20202(const xrt_core::device * _pDevice,
...
+void
+AieTopThreadFunction(const boost::property_tree::ptree& _pt)
+{
+  std::ostream &_output = std::cout;
+
+  while(!is_quit)
+  {
+    try
+    {
+#if defined _WIN32
```
Comment (rozumx) [inline src/runtime_src/core/tools/common/ReportAie.cpp:522]: Please, no OS directives.

File: src/runtime_src/core/tools/common/ReportAie.cpp [hunk 5]
```diff
@@ -502,14 +510,108 @@ ReportAie::getPropertyTree20202(const xrt_core::device * _pDevice,
...
+}
+
+void
+AieTopFunc(const boost::property_tree::ptree& _pt)
+{
+  boost::asio::io_service io_service;
+
+  is_quit = false;
+  // Create a threadmain function for printing the status
+  std::thread t_thread(AieTopThreadFunction, _pt);
```
Comment (rozumx) [inline src/runtime_src/core/tools/common/ReportAie.cpp:581]: This is a major architectural use-case change.  Reports are currently designed to take a one (1) sample reading and report the results.  This "threading" concept (e.g., equivalent to xbutil top), is is currently scheduled for 2022.1.

**Review by rozumx — CHANGES_REQUESTED** [review]
This functional is a major change from how the current code operations.   In addition, for 2022.1, we will be adding support for requests like this.

Q: Do these changes really need to be in in 2021.2?

File: src/runtime_src/core/tools/common/ReportAie.cpp (hunk 1 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportAie.cpp]: Removed.

File: src/runtime_src/core/tools/xbutil2/CMakeLists.txt (hunk 3 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/xbutil2/CMakeLists.txt]: Removed the thread. Now no thread functionality is implemented.

File: src/runtime_src/core/tools/common/ReportAie.cpp (hunk 4 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportAie.cpp]: This functionality is removed.

File: src/runtime_src/core/tools/common/ReportAie.cpp (hunk 5 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportAie.cpp]: This functionality is removed. We will take care this portion in an another story for 2022.1 https://jira.xilinx.com/browse/VITIS-3365

File: src/runtime_src/core/tools/common/ReportAie.cpp [hunk 6]
```diff
@@ -502,13 +504,26 @@ ReportAie::getPropertyTree20202(const xrt_core::device * _pDevice,
...
                        std::ostream & _output) const
 {
   boost::property_tree::ptree empty_ptree;
+  std::vector<std::string> aieCoreList;
+  bool is_less = false;
+
+  for (auto it = _elementsFilter.begin(); it != _elementsFilter.end(); ++it) {
+    if(*it == "cores") {
+      auto core_list = std::next(it);
+      if (core_list != _elementsFilter.end())
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/tools/common/ReportAie.cpp:520]: Please do some negative testing around these options.

**Review by chvamshi-xilinx — APPROVED** [review]

**Review by stsoe — APPROVED** [review]

File: src/runtime_src/core/tools/common/ReportAie.cpp [hunk 7]
```diff
@@ -537,7 +552,12 @@ ReportAie::writeReport(const xrt_core::device* /*_pDevice*/,
       _output << std::endl;
       count = 0;
       for (auto& tile : graph.get_child("tile")) {
-        _output << boost::format("Core [%2d]\n") % count++;
+	int curr_core = count++;
+	if(aieCoreList.size() && (std::find(aieCoreList.begin(), aieCoreList.end(),
```
Comment (sonals) [inline src/runtime_src/core/tools/common/ReportAie.cpp:556]: Do you need the .size() test? std::find() should be fine with empty list?

File: src/runtime_src/core/tools/common/ReportAieShim.cpp [hunk 8]
```diff
@@ -29,6 +31,7 @@
 
 
 namespace qr = xrt_core::query;
+std::vector<std::string> aieTileList;
```
Comment (sonals) [inline src/runtime_src/core/tools/common/ReportAieShim.cpp:34]: Are we adding a global variable here?

**Review by sonals — COMMENTED** [review]
Please add a single white space after `if` statement.

File: src/runtime_src/core/tools/common/ReportAie.cpp (hunk 2 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportAie.cpp]: Taken care.. Removed this..

File: src/runtime_src/core/tools/common/ReportAieShim.cpp (hunk 8 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportAieShim.cpp]: Not required to be a global variable. Moved it as a local variable .. Thanks

File: src/runtime_src/core/tools/common/ReportAie.cpp (hunk 7 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportAie.cpp:556]: Actually if aieCoreList is empty then we should fail this if condition and print all the cores information. Hence the size check is required here. 
If not empty then we will try to find the specified core..

File: src/runtime_src/core/tools/common/ReportAie.cpp (hunk 6 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportAie.cpp:520]: I have done that. With some value provided or without
 provided any  value

File: src/runtime_src/core/tools/common/ReportAie.cpp (hunk 7 above)
Comment (keryell) [reply src/runtime_src/core/tools/common/ReportAie.cpp:556]: If you are interested by emptiness, just use `aieCoreList.empty()`

File: src/runtime_src/core/tools/common/ReportAieShim.cpp [hunk 9]
```diff
@@ -21,6 +21,8 @@
 #include "core/common/device.h"
 #include <boost/optional/optional.hpp>
 #include <boost/property_tree/json_parser.hpp>
+#include <boost/algorithm/string.hpp>
```
Comment (keryell) [inline src/runtime_src/core/tools/common/ReportAieShim.cpp:24]: Sort

File: src/runtime_src/core/tools/common/ReportAieShim.cpp [hunk 10]
```diff
@@ -256,13 +258,22 @@ ReportAieShim::getPropertyTree20202(const xrt_core::device * _pDevice,
...
   _output << "AIE\n";
 
+  for (auto it = _elementsFilter.begin(); it != _elementsFilter.end(); ++it) {
+    if(*it == "tiles") {
+      auto tile_list = std::next(it);
+      if (tile_list != _elementsFilter.end())
+        boost::split(aieTileList, *tile_list, boost::is_any_of(","));
+    }
+  }
+
```
Comment (keryell) [inline src/runtime_src/core/tools/common/ReportAieShim.cpp:276]: This looks like déjà vu.
Should we have some common functions instead to process the `_elementsFilter` and pass some actions as lambdas?

File: src/runtime_src/core/tools/common/ReportAie.cpp [hunk 11]
```diff
@@ -502,13 +504,26 @@ ReportAie::getPropertyTree20202(const xrt_core::device * _pDevice,
...
+  bool is_less = false;
+
+  for (auto it = _elementsFilter.begin(); it != _elementsFilter.end(); ++it) {
+    if(*it == "cores") {
+      auto core_list = std::next(it);
+      if (core_list != _elementsFilter.end())
+        boost::split(aieCoreList, *core_list, boost::is_any_of(","));
+    }
+    if(*it == "less") {
+      is_less = true;
```
Comment (keryell) [inline src/runtime_src/core/tools/common/ReportAie.cpp:524]: Add a comment or rename to something more explicit?
I have just understood what it means by reading your super cool PR description.
The I realize that all this information from the PR description will be lost to the people reading your interesting code.
You could summarize what it does in a big design comment with motivation and example above the display function.

File: src/runtime_src/core/tools/common/ReportAie.cpp [hunk 12]
```diff
@@ -21,6 +21,8 @@
 #include "core/common/device.h"
 #include <boost/optional/optional.hpp>
 #include <boost/property_tree/json_parser.hpp>
+#include <boost/algorithm/string.hpp>
+#include <boost/algorithm/string/split.hpp>
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportAie.cpp:23]: Fixed. Thanks

File: src/runtime_src/core/tools/common/ReportAie.cpp [hunk 13]
```diff
@@ -502,13 +504,26 @@ ReportAie::getPropertyTree20202(const xrt_core::device * _pDevice,
...
-                       const std::vector<std::string>& /*_elementsFilter*/, 
+                       const boost::property_tree::ptree& _pt,
+                       const std::vector<std::string>& _elementsFilter,
                        std::ostream & _output) const
 {
   boost::property_tree::ptree empty_ptree;
+  std::vector<std::string> aieCoreList;
+  bool is_less = false;
+
+  for (auto it = _elementsFilter.begin(); it != _elementsFilter.end(); ++it) {
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportAie.cpp:518]: Added a comments.

File: src/runtime_src/core/tools/common/ReportAie.cpp (hunk 11 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportAie.cpp:530]: Added comments. Please check and share your feedback.

File: src/runtime_src/core/tools/common/ReportAieShim.cpp (hunk 9 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportAieShim.cpp:22]: Fixed. Thanks

File: src/runtime_src/core/tools/common/ReportAieShim.cpp (hunk 10 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportAieShim.cpp:279]: _elementsFilter  is not used very commonly. Adding command function will lead more code change.

File: src/runtime_src/core/tools/common/ReportAie.cpp [hunk 14]
```diff
@@ -537,7 +552,12 @@ ReportAie::writeReport(const xrt_core::device* /*_pDevice*/,
       _output << std::endl;
       count = 0;
       for (auto& tile : graph.get_child("tile")) {
-        _output << boost::format("Core [%2d]\n") % count++;
+	int curr_core = count++;
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportAie.cpp]: Fixed. Thanks

File: src/runtime_src/core/tools/common/ReportAie.cpp [hunk 15]
```diff
@@ -537,7 +558,12 @@ ReportAie::writeReport(const xrt_core::device* /*_pDevice*/,
...
       for (auto& tile : graph.get_child("tile")) {
-        _output << boost::format("Core [%2d]\n") % count++;
+        int curr_core = count++;
+        if(aieCoreList.size() && (std::find(aieCoreList.begin(), aieCoreList.end(),
+		     std::to_string(curr_core)) == aieCoreList.end()))
+          continue;
+
+        _output << boost::format("Core [%2d]\n") % curr_core;
         _output << fmt4("%d") % "Column" % tile.second.get<int>("column");
         _output << fmt4("%d") % "Row" % tile.second.get<int>("row");
```
Comment (AShivangi) [inline src/runtime_src/core/tools/common/ReportAie.cpp:568]: is there a reason "fmt4" was chosen? throughout xbutil reports, we are following indentation of 2

**Review by AShivangi — APPROVED** [review]

**Review by rozumx — APPROVED** [review]
Approval dependent on previous reviewers whom I agree with their observations.

**Pull request merged**